### ISSUE TITLE
Converting any HTML document to PDF -including FS-loaded, and empty files- takes at least 30 seconds (Windows 10)

### ISSUE THREAD

**Issue by dev-eduka** [wkhtmltopdf/wkhtmltopdf#4891]
**wkhtmltopdf version(s) affected**: all versions, tried rolling back all the way to 12.4, tried portable version, installer version, 32 bit, 64 bit...

**OS information**
Windows 10 Home x64 with latest updates

**Description**
Since recently, wkhtmltopdf is taking over 30 seconds to convert the simplest blank page.

**How to reproduce**
Create an empty .html file and put one word in it such as "test" or anything else.
Run the conversion command with the least possible number of arguments such as "wkhtmltopdf.exe c:\temp\test.html c:\temp\test.pdf"
I observed this behavior through WampServer at first, but once I started switching to command line and locally loaded files I noticed the error was still happening.

**Expected behavior**
Should be nearly instant.
Instead, wkhtmltopdf takes at least 30 seconds to process the file. At the end, the output file is correct.
The app is stuck at "Loading pages.... [100%] 1/6"
While doing that, I observed in the task manager that the wkhtmltopdf process is doing nothing. No I/O, CPU, or memory usage.
I tried literally everything, including rebooting multiple times, reinstalling versions, cleaning up, using some command arguments like caching and whatnot, disabling Windows firewall, disabling antivirus, killing all other non-essential/non-system processes, etc.

**Possible Solution**  
I couldn't figure out a solution for Windows. So I installed WSL2, Ubuntu, wkhtmltopdf (0.12.5-1 from repo) and that works correctly. But it's a challenge to get the linux command to interact with WAMP.

A video of the issue has been uploaded here: https://share.getcloudapp.com/ApuGNNwg

**Comment by cwenger** [comment]
Exact same issue here. I suspect maybe a Windows Update did it?

**Comment by dev-eduka** [comment]
Glad this is getting some traction. My colleague who's working under the same environment as I do, same software versions, etc. isn't getting the issue. Don't know what's different between us. His laptop is a solid/recent Huawei.
Mine is a Lenovo with an Intel i7 CPU
Full specs: https://prnt.sc/vvgju9
Doubt it's got anything to do with computing power though since the process is basically idle/frozen.

**Comment by PhilterPaper** [comment]
Have you checked that other file-I/O-heavy applications, such as compilers, haven't slowed down too? No network traffic with what you're running (all behind the firewall)? You've run a virus/malware check? 

You mentioned that the Task Manager shows no real activity (although I've found TM to be rather unreliable). What could be drastically slowing down local file-I/O? No sign that you have a failing HDD or something like that? Any hints as to whether it's just waiting forever to _start_ the file read, and then gulps it down quickly, or is it chugging along as slow as molasses?

**Comment by dev-eduka** [comment]
> Have you checked that other file-I/O-heavy applications, such as compilers, haven't slowed down too? No network traffic with what you're running (all behind the firewall)? You've run a virus/malware check?
> 
> You mentioned that the Task Manager shows no real activity (although I've found TM to be rather unreliable). What could be drastically slowing down local file-I/O? No sign that you have a failing HDD or something like that? Any hints as to whether it's just waiting forever to _start_ the file read, and then gulps it down quickly, or is it chugging along as slow as molasses?

I have done the following:
- killed every other process (other than system) to ensure no process was consuming cpu/ram/io/network resources
- Rebooted serveral times
- The IO is excellent in general (high performance SSD) and I have no particular issue
- Not a failing SSD from what I can tell, the laptop is new (purchased less than 6 months ago) I ran chkdsk to verify that too, no errors
- The network isn't involved as far as I'm concerned since I'm converting local files / empty files / blank documents and I get the same reaction every time. The same happens no matter what the source file looks like
- As you can see on the video the 1st step "Loading page" is almost instant. It just freezes at 100% for about 30 seconds

If there's anything I can do to help debug this please let me know.

**Comment by cwenger** [comment]
@dev-eduka Can you share more detailed Windows build information? If you go to 'check for updates' there is an 'OS build info' link. Here's mine:

> Edition	Windows 10 Pro
> Version	20H2
> Installed on	‎10/‎27/‎2020
> OS build	19042.662
> Experience	Windows Feature Experience Pack 120.2212.551.0

I can't say for sure but I think this issue started around the time when I upgraded to 20H2. Another similar system that is _not_ exhibiting this problem:

> Edition	Windows 10 Pro
> Version	1909
> Installed on	2/‎24/‎2020
> OS build	18363.1198

Unfortunately it won't let me update this system to confirm that it causes the problem.

**Comment by PhilterPaper** [comment]
Just because Windows is identical doesn't mean anything if any of the hardware is different. For instance, you could have a different graphics card whose driver (on the latest build) is sucking up all the cycles, while on the other system (same Windows build) it's nicely behaved.

**Comment by cwenger** [comment]
@PhilterPaper Agreed but I thought maybe it would be something as simple as every system with some particular Windows update has the problem. As far as I can tell, the system is running flawlessly otherwise. It was just reformatted a few months ago.

**Comment by dev-eduka** [comment]
Processor	Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz   2.30 GHz
Installed RAM	16,0 GB (15,8 GB usable)
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display
Edition	Windows 10 Home
Version	20H2
Installed on	‎31/‎08/‎2020
OS build	19042.630
Serial number	R90W6Q2B
Experience	Windows Feature Experience Pack 120.2212.31.0

Doubt this helps. I did actually update my graphics card drivers a few times recently. It's not entirely impossible that the slowdowns started occurring after this. I'm going to make an attempt using windows's safe mode.

**Comment by dev-eduka** [comment]
**TL;DR: Windows's network printing service causes this, set a local/virtual printer as default printer to fix the issue**

I have, ladies and gentlemen, figured it out.
After rebooting in safe mode (no network, no video card, etc.) the command ran correctly and gave me a warning:
c:\Program Files\wkhtmltopdf\bin>wkhtmltopdf.exe c:\temp\test.html c:\temp\test4.pdf
Loading pages (1/6)
**QWin32PrintEngine::initialize: OpenPrinter failed (The RPC server is unavailable.)
QWin32PrintEngine::initialize: OpenPrinter failed (The RPC server is unavailable.)**
Counting pages (2/6)
Resolving links (4/6)
Loading headers and footers (5/6)
Printing pages (6/6)
Done

That reminded me I have a network printer that's offline and causing annoying timeouts in Microsoft Word too when changing the margins. So I tried the following:
- "Airplane mode" in Windows: wkhtmltopdf command runs well (instant)
- Disable "Print spooler" service: wkhtmltopdf command runs well (instant)
- Turn on the network printer so it's reachable: wkhtmltopdf command runs well (a little bit slow, must somehow be due to the fact that it communicates with the network printer)
- Set another printer as default printer (you have to untick "Let Windows manage my default printer" first in the Devices & Printers window) such as the Microsoft XPS printer: command runs well (instant)

I'm going to close this now. Hope this solution helps.

**Comment by cwenger** [comment]
Yep that's it! Interestingly once I installed the manufacturer software for my default printer, the problem went away, even when it's offline--this apparently was the difference between my two similar systems. Thanks so much for solving this. What a coincidence that we both encountered this issue within a few days of each other.

**Comment by KrullBorg** [comment]
same here on gnu/linux: stopping cups and then it worked again very fast

**Comment by bezlin6mechminerz** [comment]
If we turn of the internet the process is super fast.

**Comment by Wissperwind** [comment]
Same issue here. What a connection of problems.
I use my network printer on this machine rarely. So I use the windows standard driver.
If it is the standard printer -> wkhtmltopfd is extremely slow.
If I select another printer a standard printer -> problem solved.
I now might take the effort to install the correct driver for that printer...

**Comment by PIoneer2** [comment]
2022-12 - same here on Windows 10, `wkhtmltopdf` version 0.12.6.1.
Yesterday it was fast but today - slows after step 1/6...
The idea of a remote printer was right. Today has a network printer as the default. And it started to happen.
*My solution: set a software printer as the default one ("Microsoft Print to PDF").* Now, it works as used to.

IMHO, the issue should be reopened. Due to `wkhtmltopdf`s work in a headless mode - it should not rely on the head. Especially, when it is not in use.
